integrationtest/elasticsearch/src/test/java/org/hibernate/search/elasticsearch/test/ElasticsearchIndexManagerClosingIT.java

```diff
@@ -0,0 +1,204 @@
+
+		String indexUrlRegex = "^/" + Entity.INDEX_NAME + "$";
+		String indexAndTypeMappingUrlRegexPrefix = "^/" + Entity.INDEX_NAME + "/"
+				+ URLEncodedString.fromString( Entity.class.getName() ).encoded;
+
+		// ----------------------------------------------
+		// Expected Elasticsearch request sequence
+		// 1. Init the index
+		expect( stubClient.submit( requestMatching( "HEAD", indexUrlRegex ) ) )
+				.andReturn( CompletableFuture.completedFuture( notFoundResponse() ) );
+		expect( stubClient.submit( requestMatching( "PUT", indexUrlRegex ) ) )
+				.andReturn( CompletableFuture.completedFuture( okResponse() ) );
+		expect( stubClient.submit( requestMatching(
+						"GET",
+						"^/_cluster/health/" + Entity.INDEX_NAME + "$"
+				) ) )
+				.andReturn( CompletableFuture.completedFuture( okResponse() ) );
+		expect( stubClient.submit( requestMatching(
+						"PUT",
+						indexAndTypeMappingUrlRegexPrefix + "/_mapping$"
+				) ) )
+				.andReturn( CompletableFuture.completedFuture( okResponse() ) );
+		// 2. Index a document
+		expect( stubClient.submit( requestMatching( "POST", "^/_bulk$" ) ) )
+				.andReturn( blockedResponseFuture );
+		// 3. And only *then* probe for index existence and delete the index
+		expect( stubClient.submit( requestMatching( "HEAD", indexUrlRegex ) ) )
+				.andReturn( CompletableFuture.completedFuture( okResponse() ) );
+		expect( stubClient.submit( requestMatching( "DELETE", indexUrlRegex ) ) )
+				.andReturn( CompletableFuture.completedFuture( okResponse() ) );
+		// End of expected Elasticsearch request sequence
+		// ----------------------------------------------
+
+		stubClient.close();
+
+		replay( stubClientFactory, stubClient );
+
+		SearchConfigurationForTest configuration = new SearchConfigurationForTest()
+				.addClass( Entity.class )
+				.addProvidedService( ElasticsearchClientFactory.class, stubClientFactory )
+				.addProperty(
+						// This is absolutely mandatory to reproduce the bug
+						"hibernate.search.default." + ElasticsearchEnvironment.INDEX_SCHEMA_MANAGEMENT_STRATEGY,
+						IndexSchemaManagementStrategy.DROP_AND_CREATE_AND_DROP.getExternalName()
+				);
+
+
+		ExtendedSearchIntegrator searchIntegrator = searchIntegratorResource.create( configuration );
+
+		Entity entity = new Entity( 0 );
+		AddLuceneWork work = searchIntegrator.getIndexBinding( Entity.TYPE_IDENTIFIER )
+				.getDocumentBuilder().createAddWork(
+						null, Entity.TYPE_IDENTIFIER, entity, 0, "0",
+						SimpleInitializer.INSTANCE, new ContextualExceptionBridgeHelper()
+				);
+		// Important: we actually want to perform a stream, async operation here.
+		searchIntegrator.getIndexManager( Entity.INDEX_NAME )
+				.performStreamOperation( work, null, true );
+
+		// Schedule the work to be actually executed a long time from now,
+		// so that we can be sure the closing process waited for it to execute
+		ForkJoinPool.commonPool().submit( () -> {
+			try {
+				Thread.sleep( 1_000 );
```

Comment:
Could you avoid waiting for an arbitrary amount of time? It's almost never a good idea in tests: either we end up with very slow tests as they get littered with "wait" operations, or often we're not even sure if the delay is going to be enough for the machines actually running the tests.

In this case you could just have a barrier, which is broken right after the searchIntegrator.close() ? Does it even have to be a different thread to mark the "complete" phase?

Reply:
The change needs to happen sometime *during* the call to `searchIntegrator.close()`, because if all goes well the thread closing the search integrator is actually waiting for the change to happen. If the change happens before the call to `close()`, we won't be testing that the wait occurs, and if it happens after the call to `close()`, the call to `close()` will never terminate (it will wait forever).

So I definitely need another thread, and I'm afraid I can't use a barrier, because the barrier would actually need to be broken sometime *during* the `searchIntegrator.close()`... I would need some extension point inside the index manager allowing me to break the barrier just after the index manager started waiting on pending threads...

Please, don't answer "use byteman" xD

Reply:
use ASM then! xD

Reply:
Can you possibly invert the logic? do an "asynch close" which is blocking forever - unless the event is done.
Something along these lines should allow you to avoid picking an arbitrary sleep time.

Reply:
my bad I misunderstood the intent. We agreed on adding a reminder comment and merge.